examples/matrice-factorisation/Main.hs

```diff
@@ -0,0 +1,65 @@
+{-# LANGUAGE BlockArguments #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE RecordWildCards #-}
+
+module Main where
+
+import Control.Monad (foldM, when)
+import Torch.Autograd (grad, makeIndependent, toDependent)
+import Torch.DType (DType (Float))
+import Torch.Functional (matmul, mse_loss)
+import Torch.NN
+  ( Parameter,
+    Randomizable,
+    sample,
+    sgd,
+  )
+import Torch.Tensor (Tensor, asTensor)
+import Torch.TensorFactories (randn')
+
+data MatrixFactSpec = MatrixFactSpec {common_dim :: Int, extra_dim :: Int} deriving (Show, Eq)
+
+data MatrixFact = MatrixFact {u :: Parameter, v :: Parameter}
+
+instance Randomizable MatrixFactSpec MatrixFact where
+  sample (MatrixFactSpec n k) = do
+    u <- makeIndependent =<< randn' [n, k]
+    v <- makeIndependent =<< randn' [k, n]
+    pure $ MatrixFact u v
+
+flattenParameters :: MatrixFact -> [Parameter]
+flattenParameters (MatrixFact u v) = [u, v]
+
+replaceParameters :: MatrixFact -> [Parameter] -> MatrixFact
+replaceParameters state params_list = MatrixFact u v
+  where
+    u = params_list !! 0
+    v = params_list !! 1
+
+mulMF :: MatrixFact -> Tensor
+mulMF (MatrixFact u v) = matmul u' v'
+  where
+    u' = toDependent u
+    v' = toDependent v
+
+lossMF :: MatrixFact -> Tensor
+lossMF mf = mse_loss (mulMF mf) r
+  where
+    r = asTensor ([[1, 2, 3], [4, 5, 6], [7, 8, 9]] :: [[Float]])
+
+num_iters = 2000
+
+main :: IO ()
+main = do
+  init <- sample $ (MatrixFactSpec 3 4)
```

Comment:
you *may* want to seed the random number generator so that the run is reproducible, see https://github.com/hasktorch/hasktorch/blob/master/examples/static-mnist-mlp/Main.hs#L143 for an example of that.
we have been working towards exposing the RNG so that one can pass it around as a value, but that's not ready yet.

Reply:
another thought: it seems the number of latent factors `k` is 4 and thus larger than the rank of your matrix `r` above. this is probably not a good choice, since typically you want to use use fewer factors than `rank r`.

Reply:
furthermore, `r` is fully populated. I would suggest that, maybe in another PR, you look at the more practically relevant case in which `r` is sparse. you can then split `r` in a training and validation set, where the validation task is to reconstruct the missing values.

Reply:
Your points are great.
* I will make the seed fixed such that it becomes reproducible.
* I guess the optimal value for `k` will be `rank(r)` (not sure though).
* Sparse `r` will be dfinitely more interesting and closer to the applications.

Reply:
1. Great, thank you!
2. In the general case, in which `r` is sparse, the optimal value will probably be smaller than `rank r`. I would treat `k` as a hyper parameter. You can leave it as it though for now.
3. Yes, that’s no problem. I would have suggested that as well.

Cheers and thank you!